### DISCUSSION TITLE
chipkit core pins map

### DISCUSSION THREAD

**Discussion by G2N2222** [chipKIT32/chipKIT-core#468] [General]
Hi, i juste start whith chipkit, insatling on vscode and platformio pluging.
I make tests on UBW32 board( pic32mx460f512 version).
For start, i want a simple blink example. It compil fine, load fine, but nothing blink??
So....in .platformioconfig there no ubw32 460 board environement, i choose mx4pro who have the same chip, but not the same buidin leds.
On ubw32 460 leds are connected to E0,1,2,3.
On mx4pro documentation, E0....are maped to 98..pin. i try, nothing blink. After ride up from pinmode fonction, i find pin map on arduino.h file. PortE and pin 0 maped to 1number pin.i try and now it blinks!
Arduino.h pin map seems to refert to mega boad.
How can i know the good pin map for my board or for mx4pro similar board
Thanxs.

**Comment by majenkotech** [comment]
I don't know much about platformio, but looking at their database  there is the UBW32 MX460 variant available: https://docs.platformio.org/en/latest/boards/microchippic32/ubw32_mx460.html

**Comment by G2N2222** [comment]
thank you. I have already seen this site. but I can't find a doc to match the mcu pins (E0) and arduino pins (0,1,2 ..) used in a program and why chipkit refers to an arduino.h file which seems to correspond to a mega2560?

**Comment by EmbeddedMan** [comment]
Yeah, my apologies for that. The MX460 version of UBW32 was discontinued a long time before any support was added to chipKIT for UBW32, and even then, the support that is there is extremely minimal and not obvious to a new user. It's been many years since I've looked at the UBW32 support in chipKIT - I'll dig into it this weekend and report back here everything I find.

**Reply by majenkotech** [reply]
It looks like the UBW32 MX4 uses the generic "100 pin" variant.  For this the pins are numbered by the port names, not the board pin numbers: PORTA bit 0 is pin 0. PORTA bit 1 is pin 1, etc. PORTB bit 0 is pin 16, PORTC bit 0 is 32, etc.  So E0 would be (A=0, B=1, C=2, D=3, E=4, etc) 16 * E + 0, so pin 64. 

The files to look at for the pin maping are `Board_Defs.h` and `Board_Data.c`, wherever PlatformIO puts those...

**Comment by G2N2222** [comment]
EmbeddedMan: yes it's old board, but it's only i have. that's why i try to use mx4 pro board environement.

majenkotech : yes it's my reasoning too, but the only working code is if i use digital pin 1:

`#define LED_BUILTIN 1

void setup()
{
// initialize LED digital pin as an output.
pinMode(LED_BUILTIN, OUTPUT);
...........
..........
`

it seems like the pin mapping was compil from pins_arduino.h located at:
C:\Users......platformio\packages\framework-arduino-avr\cores\arduino

.....packages\framework-arduino-avr....!! .that it choc me , why not from:
C:\Users.....platformio\packages\framework-arduinomicrochippic32\cores\pic32 and
C:\Users......platformio\packages\framework-arduinomicrochippic32\variants\UBW_MX4\Board_Data.c ?

i try with arduino IDE and UECIDE i have same probléme........ maybe a path probleme?

**Comment by G2N2222** [comment]
OK now it's work!!!!

While exploring the chipkitcore files, I noticed that the mx4-pro and the ubw460 did not have the same board_data.c and .h file
I solved the problem by creating a ubw460 environment in platformio.ini on the ubw790 model.
which had the effect, I think, of correctly mapping the board_data file, instead of using an arduino mega file by default

**Reply by majenkotech** [reply]
That's great. Do you fancy back-porting that new board definition to the chipKIT core repo with a pull request?

**Comment by G2N2222** [comment]
No, I don't think this is necessary. I only added 4 lines to a file only used by platformio.et for arduino, after a cleanup and and clean reinstall in another directory, it works. Again thanks for your suggestions which prompted me to take a close look at the structure of chipkitcore.
